source/language/openpulse.rst

```diff
@@ -0,0 +1,609 @@
+type which is responsible for tracking two properties:
+
+- Tracking time appropriately so programs do not need to deal in absolute time or with the
+  bookkeeping of advancing time in a sequence of pulses.
+- Tracking phase by producing a complex value given an input time (i.e. via the mathematical
+  relationship :math:`e^{i\left(2\pi f t + \theta\right)}`,  where `f` is frequency and
+  :math:`\theta` is phase). One motivation for keeping track of phase is to allow pulses to be
+  defined in the rotating frame with the effect being an equivalent application in the lab
+  frame (i.e. with the carrier supplied by the ``frame``). Another motivation is to more naturally
+  implement a "virtual Z-gate", which does not require a physical pulse but rather shifts the phase
+  of all future pulses on that frame.
+
+The frame is composed of three parts:
+
+1. A frequency ``frequency`` of type ``float``.
+2. A phase ``phase`` of type ``angle``.
+3. A time which is manipulated implicitly and cannot be modified other
+   than through existing timing instructions like ``delay`` and ``barrier``. The time increment
+   is determined by the channel on which the frame is played (see :ref:`Timing` section).
+
+The ``frame`` type is a virtual resource and the exact precision of these parameters is
+hardware specific. It is thus up to the compiler to choose how to implement the required
+transformations to physical resources in hardware (e.g. mapping multiple frames to a
+single NCO).
+
+Frame Construction
+~~~~~~~~~~~~~~~~~~
+
+Frames can be constructed using the ``newframe`` command e.g.
+
+.. code-block:: javascript
+
+  frame driveframe = newframe(5e9, 0.0); // newframe(float[size] frequency, angle[size] phase)
+
+would construct a frame with frequency `f`  5 GHz and phase :math:`\theta` 0.0. When
+instantiated, the frame time starts at 0. ``frame``s can also be copied using the
+``copyframe`` command
+
+.. code-block:: javascript
+
+  frame driveframe1 = newframe(5e9, 0.0);
+  frame driveframe2 = copyframe(driveframe1);
+  driveframe2.phase = driveframe1.phase + pi/2;
+
+This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` with the same
+time as `driveframe1`.
+
+Frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase) tuple of a frame can be manipulated throughout program
+by referencing ``.frequency``, and ``.phase``. Operations must be
+appropriate for the respective type, ``float`` for frequency and ``angle`` for
+phase. Again, the exact precision of these calculations is hardware specific.
+
+Here's an example of manipulating the phase to calibrate an ``rz`` gate on a frame called
+``driveframe``:
+
+.. code-block:: javascript
+
+   // Example 1: Shift phase of the "drive" frame by pi/4, eg. an rz gate with angle -pi/4
+   cal {
+     driveframe.phase += pi/4;
+   }
+
+   // Example 2: Define a calibration for the rz gate on all 8 physical qubits
+   cal {
+     frame[8] rz_frames;
```

Comment:
To be revisited when we intro `angle` syntax, but I'd prefer that `frame` not take a width, and you instead re-write this line as `angle[frame, 8] rz_frames`.

Reply:
I think you mean `array[frame, 8] rz_frames`, but yes I agree.

Reply:
Updated array frames here e858e5f. Left a comment noting we're stilling trying to resolve this.